APID 2.9.0: Copying info from the reviewed release notes into the topic about AMF validations (#108)
mulesoft/docs-design-center: modules/ROOT/pages/design-common-problems-raml-10.adoc

<<Back to the top>>

=== Types defined as `nil` in payloads are not valid

> You cannot use a type in a payload if its data type is `nil`.

Reviewer: ```suggestion
You cannot use a `nil` data type in a payload.
```
...to avoid any possible confusion with the payload having a data type....

==== Example 1

In this example, the value of `type` must be `string`, not `nil`:

```
#%RAML 1.0
title: test

types:
  MyType:
    type: nil
    example: “test”
```

==== Example 2

> In this example, the JSON schema must define `MyType` as a `string` type. However, the schema is empty.

Reviewer: ```suggestion
In this example, the JSON schema must define `MyType` as a `string` type. However, the schema is empty:
```

<<Back to the top>>

[.resolution\*invalid-type-inheritance#resolution\*invalid-type-inheritance]
== Invalid Type Inheritance

> The errors in this section both involve incorrect handling of type inheritance.

Reviewer: Avoid self-reference to topics, sections, documents...:

```suggestion
Two common occurrences that cause type-inheritance errors are incorrectly merging two types and defining a weaker constraint in a subtype.

> === Merging two types incorrectly

Reviewer: ```suggestion
=== Merging Two Types Incorrectly
```

<<Back to the top>>

=== Defining a Weaker Constraint in a Subtype

> A subtype that defines a weaker constraint than one defined by its parent is invalid.

Reviewer: Should we use "child" with "parent"?
Or "subtype" with "primary type"?
parallelism

> In this example, the subtype MyType2 has the optional property `name`, defined as a string, even though the parent type MyType1 already defines its `name` property as required.

Reviewer: ```suggestion
In this example, the subtype MyType2 has the optional property `name`, defined as a string, even though the parent type MyType1 already defines its `name` property as required:
```

|====

<<Back to the top>>

== Using a Forward Slash in a URI Parameter

A URI parameter that contains a forward slash (/) is invalid.

> In this example, the `example` for the URI parameter `transaction-Date` would cause the parser to display an error message.

Reviewer: ```suggestion
In this example, the `example` for the URI parameter `transaction-Date` causes the parser to display an error message:
```

```
#%RAML 1.0
title: test

/endpoint:
  uriParameters:
    transaction-Date:
      example: 01/02/2018
```

The error message in cases like this is `Value '01/02/2018' of uri parameter must not contain '/' character.`

<<Back to the top>>

== Incorrect Formats for `datetime` Types

Formats for `datetime` types that do not use "rfc3339" or "rfc2616" are invalid.

> The following example would cause the parser to display an error message.

Reviewer: ```suggestion
The following example causes the parser to display an error message:
```
